app/shared-components/PromoModal/PromoModal.jsx

```diff
@@ -0,0 +1,201 @@
+	 * @param modal			'insights' or 'premium'
+	 * @private
+	 * Handle clicks on the link to turn off promos in the promo modals
+	 */
+	_handlePromoGoAwayClick = (modal) => {
+		this.props.actions.togglePromoModal();
+
+		sendMessage('promoModals.turnOffPromos', {});
+
+		switch (modal) {
+			case INSIGHTS:
+				sendMessage('ping', 'promo_modals_decline_insights_upgrade');
+				break;
+			case PLUS:
+				sendMessage('ping', 'promo_modals_decline_plus_upgrade');
+				break;
+			case PREMIUM:
+				sendMessage('ping', 'promo_modals_decline_insights_upgrade');
+				break;
+			default:
+				break;
+		}
+
+		this.props.actions.showNotification({
+			classes: 'warning',
+			reload: false,
+			text: t('promos_turned_off_notification'),
+		});
+	};
+
+	/**
+	 * @private
+	 * Handle clicks on the download buttons
+	 */
+	_handlePromoTryProductClick = (product, utm_campaign) => {
+		this.props.actions.togglePromoModal();
+
+		let url;
+		switch (product) {
+			case PLUS:
+				// Should we send a plus ping here?
+				// sendMessage('ping', 'promo_modals_plus_upgrade_cta');
+				url = `https://checkout.${DOMAIN}.com/plus?utm_source=gbe&utm_campaign=${utm_campaign}`;
+				break;
+			case PREMIUM:
+				url = `https://ghostery.com/thanks-for-downloading-midnight?utm_source=gbe&utm_campaign=${utm_campaign}`;
+				break;
+			case INSIGHTS:
+				sendMessage('ping', 'promo_modals_insights_upgrade_cta');
+				url = `https://checkout.${DOMAIN}.com/insights?utm_source=gbe&utm_campaign=${utm_campaign}`;
+				break;
+			default:
+		}
+
+		sendMessage('openNewTab', {
+			url,
+			become_active: true,
+		});
+	}
+
+	/**
+	 * @private
+	 * Handle clicks on sign in links in promo modals
+	 */
+	_handlePromoSignInClick = () => {
+		this.props.actions.togglePromoModal();
+		history.push({
+			pathname: '/login',
+		});
+	};
+
+	_handlePromoXClick = (type) => {
+		this.props.actions.togglePromoModal();
+
+		if (type === INSIGHTS) {
+			sendMessage('ping', 'promo_modals_decline_insights_upgrade');
+		} else if (type === PLUS) {
+			sendMessage('ping', 'promo_modals_decline_plus_upgrade');
+		}
+	}
+
+	_renderXButton = (type) => {
+		const XButtonClass = ClassNames({ PlusPromoModal__exitButton: type === PLUS });
+
+		return (
+			<ModalExitButton
+				className={XButtonClass}
+				toggleModal={() => this._handlePromoXClick(type)}
+			/>
+		);
+	};
+
+	renderModalContent() {
+		const { type, loggedIn } = this.props;
+		switch (type) {
+			case INSIGHTS:
+				return (
+					<InsightsPromoModalContent
+						handleGoAwayClick={() => this._handlePromoGoAwayClick(INSIGHTS)}
+						handleTryInsightsClick={() => this._handlePromoTryProductClick(INSIGHTS, 'in_app_upgrade')}
+						handleSignInClick={this._handlePromoSignInClick}
+						{...this.props}
+					/>
+				);
+			case PLUS:
+				return (
+					<PlusPromoModalContent
+						handleGoAwayClick={() => this._handlePromoGoAwayClick(PLUS)}
+						handleTryPlusClick={() => this._handlePromoTryProductClick(PLUS, 'in_app_spring2020')}
```

Comment:
@Eden12345 @benstrumeyer It would be very good if we could avoid hardcoding a specific campaign into this generic promo modal rendering code. As is, we'll have to do a whole release if we want to end this campaign or run a different one. We should be fetching the campaign identifier(s) from a remote url that Teresa & co can update independently. If we think that's outside the scope of this release, I think we should make a ticket for it and try to get it into the next release.

Reply:
I pushed for this idea in the extension sprint meeting and it's my understanding that there's a Firefox clause preventing us from hooking up the extension to a remote URL that updates UTM params on the fly. We can sidebar this with chris since teresa was very interested in the idea.

Reply:
Ah, dang. Sounds like that Firefox policy is a dealbreaker. Too bad.